opendmarc/opendmarc.c

```diff
@@ -2509,8 +2520,54 @@ mlfi_eom(SMFICTX *ctx)
 			}
 			else if (ar.ares_result[c].result_method == ARES_METHOD_ARC)
 			{
-				if (ar.ares_result[c].result_result == ARES_RESULT_PASS)
-					dfc->mctx_arcpass = TRUE;
+				limit_arc++;
+				/*
+				** If we already came across a trusted A-R header with arc=pass
+				** we need to fail.
+				*/
+				if (ar.ares_result[c].result_result == ARES_RESULT_PASS && limit_arc == 1)
+				{
+					dfc->mctx_arcpass = ARES_RESULT_PASS;
+				}
+				else
+				{
+					dfc->mctx_arcpass = ARES_RESULT_FAIL;
+				}
+
+				/*
+				** Check arc status against whitelist policy
+				*/
+				if (dfc->mctx_arcpass == ARES_RESULT_PASS && g_hash_table_size(domain_whitelist_hash) > 0)
+				{
+					u_char *arcchain = NULL, *arcdomain;
+					int arcchainlen = 0, arcchainitempass = 0;
+
+					for (pc = 0;
+						pc < ar.ares_result[c].result_props;
+						pc++)
+					{
+						if (ar.ares_result[c].result_ptype[pc] == ARES_PTYPE_ARCCHAIN)
+						{
+							arcchain = ar.ares_result[c].result_value[pc];
+						}
+					}
+					if (arcchain != NULL)
+					{
+						arcchainlen = dmarcf_mkarray(arcchain, ":",
+						                                     &dfc->mctx_arcchain);
+						for (pc = 0; dfc->mctx_arcchain[pc] != NULL; pc++)
+						{
+							arcdomain = (u_char *)g_utf8_strdown(dfc->mctx_arcchain[pc],
+							                                     strlen(dfc->mctx_arcchain[pc]));
+							g_hash_table_contains(domain_whitelist_hash,
+							                      arcdomain) && arcchainitempass++;
```

Comment:
WAT

Reply:
Revisit. This is a logical operation. It should be, it should be a serial op.

Reply:
L2528 Is this logic correct? We need to pass on cv=none if we are the last recipient.